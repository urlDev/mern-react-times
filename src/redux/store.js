import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));