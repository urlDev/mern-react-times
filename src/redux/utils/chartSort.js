const chartSort = (prevState, newState) => {
    const tempState = [...prevState, [...newState]];

    let sortedState = [...tempState];

    sortedState = sortedState
        .sort((a, b) => a[1] - b[1])
        .map((stock) => stock.slice(0, 1));

    return sortedState;
};

export default chartSort;