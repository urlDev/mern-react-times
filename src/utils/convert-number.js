const convertNumber = (labelValue) => {
  // Nine Zeroes for Billions
  return Math.abs(Number(labelValue)) >= 1.0e9
    ? (Math.abs(Number(labelValue)) / 1.0e9).toFixed(2) + "B"
    : // Six Zeroes for Millions
    Math.abs(Number(labelValue)) >= 1.0e6
    ? (Math.abs(Number(labelValue)) / 1.0e6).toFixed(2) + "M"
    : // Three Zeroes for Thousands
    Math.abs(Number(labelValue)) >= 1.0e3
    ? (Math.abs(Number(labelValue)) / 1.0e3).toFixed(2) + "K"
    : Math.abs(Number(labelValue));
};

export default convertNumber;
