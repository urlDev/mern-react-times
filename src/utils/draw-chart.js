// Reusable function to draw historical price chart in different components
const drawChart = ({ width, reDraw, chartData, svg, d3, index }) => {
  // reDraw => Charts in market details
  // because they are drawn according to screen sizes

  let height;
  let padding;
  let chartWidth;
  let dates;

  // reDraw is for when marketDetails page is shown
  // Chart is constantly redrawn then so thats why I made a separation
  if (reDraw) {
    // changing chart width according to screen width
    chartWidth = width < 1400 ? (width < 700 ? width - 30 : width - 200) : 550;
    height = width < 1400 ? chartWidth * 0.65 : 345;
    padding = 50;
  } else {
    // when its used in market components, width and height
    chartWidth = 100;
    height = 70;
    padding = 10;
  }

  svg.attr("width", chartWidth).attr("height", height);

  if (reDraw) {
    dates = chartData.map((value) => new Date(value.date));
  } else {
    // In market components, there are more than one chart
    // So I get their indexes from the component
    dates = chartData[index].map((value) => new Date(value.date));
  }

  // xScale
  const xAxisScale = d3
    .scaleTime()
    .domain([d3.min(dates), d3.max(dates)])
    .range(
      // Depending if its reDrawn or width changes,
      // In smaller screens, I don't put any padding
      reDraw
        ? width < 500
          ? [0, chartWidth]
          : [padding, chartWidth - padding]
        : [padding, chartWidth - padding]
    );

  // yScale
  const yAxisScale = d3
    .scaleLinear()
    .domain([
      // Chart is drawn from close data on each dataset
      // So in here I am checking where the function is being used
      d3.min(reDraw ? chartData : chartData[index], (data) => data.close),
      d3.max(reDraw ? chartData : chartData[index], (data) => data.close),
    ])
    .range([height - padding, padding]);

  // Charts in market cards don't use axes
  if (reDraw) {
    //   create x Axes
    const xAxes = d3
      .axisBottom(xAxisScale)
      .ticks(width < 700 ? d3.timeDay.every(10) : 5)
      .tickFormat((date) => d3.timeFormat("%b %d, %I:%M")(date));

    // draw x axes
    width > 500 &&
      svg
        .append("g")
        .call(xAxes)
        .attr("transform", `translate(0,${height - padding})`);

    // create y axes
    const yAxes = d3.axisLeft(yAxisScale).tickFormat((d) => `$ ${d}`);

    // draw y axes
    width > 500 &&
      svg.append("g").call(yAxes).attr("transform", `translate(${padding},0)`);
  }

  // create line
  const line = d3
    .line()
    .x((d) => xAxisScale(new Date(d.date)))
    .y((d) => yAxisScale(d.close));

  // draw historical price path
  const path = svg
    .append("path")
    .data(reDraw ? [chartData] : [chartData[index]])
    .style("fill", " none")
    .attr("stroke", reDraw ? "black" : "white")
    .attr("stroke-width", reDraw ? "1.5" : "1.2")
    .attr("d", line);

  // get the length of the path for line transition
  const pathLength = path.node().getTotalLength();

  // transition for the line
  const transitionPath = d3.transition().ease(d3.easeSin).duration(2000);

  // setting pathsLength for strokes attrs and using transition
  // I made it minus pathLength because it was drawing it backwards
  path
    .attr("stroke-dashoffset", -pathLength)
    .attr("stroke-dasharray", pathLength)
    .transition(transitionPath)
    .attr("stroke-dashoffset", 0);

  path.interrupt("stroke-dashoffset");
  path.interrupt("stroke-dasharray");
};

export default drawChart;
