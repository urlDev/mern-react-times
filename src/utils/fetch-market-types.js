import { changeMarketType, changeMarketName } from "../redux/actions/chart";

const fetchMarketTypes = (dispatch, market) => {
  //  One api endpoint can be used for all market quotes
  //  So I first determine the market type
  //  Then, checking which market type client has clicked on
  //  According to market, I pre-determined tickers/symbols
  //  With a switch function, I dispatch the related market types
  //  desired tickers to redux store.
  //  So with all this, I can use one fetch function for all market kinds,
  //  Therefore, less fetch calls, less code.
  //  More comments as it seemsüòÅ
  const Indexes = "%5EGSPC,%5ERUA,%5EDJI,%5ENDX,%5EN225,%5EFTSE";
  const Crypto = "BTCUSD,LTCUSD,XLMUSD,BCNUSD,ETHUSD,ETCUSD";
  const Forex = "EURUSD,USDJPY,GBPUSD,EURGBP,EURJPY,GBPJPY";
  const Stocks = "AAPL,FB,GOOG,TSLA,NFLX,AMZN";
  const Commodities = "ZGUSD,CLUSD,HGUSD,SIUSD,PLUSD,BZUSD";

  switch (market) {
    case "Indexes":
      return [
        dispatch(changeMarketType(Indexes)),
        dispatch(changeMarketName("Indexes")),
      ];
    case "Crypto":
      return [
        dispatch(changeMarketType(Crypto)),
        dispatch(changeMarketName("Crypto")),
      ];
    case "Forex":
      return [
        dispatch(changeMarketType(Forex)),
        dispatch(changeMarketName("Forex")),
      ];
    case "Stocks":
      return [
        dispatch(changeMarketType(Stocks)),
        dispatch(changeMarketName("Stocks")),
      ];
    case "Commodities":
      return [
        dispatch(changeMarketType(Commodities)),
        dispatch(changeMarketName("Commodities")),
      ];
    default:
      return Indexes;
  }
};

export default fetchMarketTypes;
