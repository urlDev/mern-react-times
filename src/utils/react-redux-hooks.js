// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
