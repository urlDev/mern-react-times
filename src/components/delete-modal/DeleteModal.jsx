import React from "react";
import { useDispatch } from "utils/react-redux-hooks";

import { deleteModalClose, fetchDeleteUser } from "redux/actions/user";
import {
  DeleteButtonContainer,
  DeleteModalContainer,
  DeleteTextAndButtons,
  ModalContainerAllPage,
} from "./DeleteModal.styles";

const DeleteModal = () => {
  const [scroll, setScroll] = React.useState(window.scrollY);
  const dispatch = useDispatch();

  const handleClick = () => {
    dispatch(fetchDeleteUser());
    dispatch(deleteModalClose());
  };

  const updateScroll = () => {
    setScroll(window.scrollY);
  };

  // When user views the delete modal in small screens,
  // there was a bug related to where the user views it.
  // Because the top value was 0, it always put the modal to top.
  // And even though bottom was 0 too, modal opened regarding to viewport height.
  // So, by adding scroll, I could make the modal open where the user is.
  // This is the js solution.
  // In CSS, I can make it happen with only position:fixed
  // and voila!
  React.useEffect(() => {
    document.addEventListener("scroll", updateScroll);
    updateScroll();
    return () => {
      document.removeEventListener("scroll", updateScroll);
    };
  }, []);

  return (
    <ModalContainerAllPage onClick={() => dispatch(deleteModalClose())}>
      <DeleteModalContainer>
        <DeleteTextAndButtons>
          <h1>Sure about this?</h1>
          <DeleteButtonContainer>
            <button onClick={() => dispatch(deleteModalClose())}>Cancel</button>
            <button onClick={() => handleClick()}>Delete</button>
          </DeleteButtonContainer>
        </DeleteTextAndButtons>
      </DeleteModalContainer>
    </ModalContainerAllPage>
  );
};

export default DeleteModal;
