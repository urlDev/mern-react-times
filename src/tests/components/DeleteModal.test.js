import React, { useState as useStateMock } from "react";
import { shallow } from "enzyme";
import moxios from "moxios";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";

import {
  DELETE_MODAL_CLOSE,
  DELETE_USER,
  fetchDeleteUser,
} from "redux/actions/user";

import DeleteModal from "components/delete-modal/DeleteModal";

describe("Testing DeleteModal", () => {
  let wrapper;
  let store;
  let setState;
  let useStateMock;
  let reactMock;

  jest.mock("react", () => ({
    ...jest.requireActual("react"),
    useState: jest.fn(),
  }));

  beforeEach(() => {
    store = mockStore({});

    setState = jest.fn();
    useStateMock = (initState = global.scrollY) => [initState, setState];

    jest.spyOn(React, "useState").mockImplementation(useStateMock);

    jest.spyOn(React, "useEffect").mockImplementation((f) => f());

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    moxios.install();

    wrapper = shallow(<DeleteModal store={store} />);
  });

  afterEach(() => {
    moxios.uninstall();
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  test("Should render the component", () => {
    expect(wrapper).toHaveLength(1);
    expect(wrapper.find("button").first().text()).toEqual("Cancel");
  });
  test("Should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Should close modal on modal click", () => {
    wrapper.simulate("click");

    const actions = store.getActions();
    expect(actions).toEqual([{ type: DELETE_MODAL_CLOSE }]);
  });
  test("Should close modal on cancel button click", () => {
    wrapper.find("button").first().simulate("click");

    const actions = store.getActions();
    expect(actions).toEqual([{ type: DELETE_MODAL_CLOSE }]);
  });
  test("Should delete user on delete button click", async () => {
    wrapper.find("button").last().simulate("click");

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
      });
    });

    await store.dispatch(fetchDeleteUser());

    const actions = store.getActions();
    expect(actions).toEqual([
      { type: DELETE_MODAL_CLOSE },
      {
        type: "@@router/CALL_HISTORY_METHOD",
        payload: { method: "push", args: ["/"] },
      },
      { type: DELETE_USER },
    ]);
  });
  test("Should set scroll value", () => {
    global.scrollY = 1000;

    global.dispatchEvent(new Event("scroll"));

    // First one gets called in useEffect
    expect(setState).toHaveBeenCalledTimes(1);

    // Second one is for useEffect return
    global.removeEventListener(new Event("scroll"), setState(global.scrollY));

    expect(setState).toHaveBeenCalledTimes(2);
    expect(setState).toHaveBeenCalledWith(global.scrollY);
  });
});
