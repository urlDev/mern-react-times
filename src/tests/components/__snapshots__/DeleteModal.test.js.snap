// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing DeleteModal Should match snapshot 1`] = `
<styled.div
  onClick={[Function]}
  top={0}
>
  <styled.div>
    <styled.div>
      <h1>
        Sure about this?
      </h1>
      <styled.div>
        <button
          onClick={[Function]}
        >
          Cancel
        </button>
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </styled.div>
    </styled.div>
  </styled.div>
</styled.div>
`;
